geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIterator.scala

```diff
@@ -0,0 +1,289 @@
+   * @return
+   */
+  private def sortInMemory(
+      head: SimpleFeature,
+      tail: CloseableIterator[SimpleFeature],
+      ordering: Ordering[SimpleFeature],
+      closed: AtomicBoolean): CloseableIterator[SimpleFeature] = {
+    // use ArrayList for sort-in-place of the underlying array
+    val list = new java.util.ArrayList[SimpleFeature](100)
+    list.add(head)
+    // note: we already ensured tail.hasNext == true before calling this method
+    do { list.add(tail.next()) } while (tail.hasNext && !closed.get)
+    if (closed.get) {
+      CloseableIterator(Iterator.empty, tail.close())
+    } else {
+      list.sort(ordering)
+      CloseableIterator(list.iterator.asScala, tail.close())
+    }
+  }
+
+  /**
+   * Sorts the iterator in memory, but spills over to disk if the size in memory exceeds the specified threshold
+   *
+   * @param head head of iterator
+   * @param tail tail of iterator
+   * @param ordering ordering
+   * @param closed checked to determine if we should stop sorting and return early
+   * @param threshold max memory to use before spilling to disk
+   * @return
+   */
+  private def sortWithSpillover(
+      head: SimpleFeature,
+      tail: CloseableIterator[SimpleFeature],
+      ordering: Ordering[SimpleFeature],
+      closed: AtomicBoolean,
+      threshold: Long): CloseableIterator[SimpleFeature] = {
+
+    // grouping ID for tmp files, doesn't have to be unique
+    lazy val id = {
+      val name = head.getFeatureType.getTypeName.replaceAll("[^A-Za-z0-9_-]", "").take(20)
+      f"$name-${Random.nextInt(10000)}%04d"
+    }
+
+    val sizable: FeatureIsSizable[SimpleFeature] = head match {
+      case _: SimpleFeature with Sizable =>
+        SizableFeatureIsSizable.asInstanceOf[FeatureIsSizable[SimpleFeature]]
+      case _ =>
+        // currently, all of our uses of this class pass in Sizable features... this means we missed one
+        logger.warn(
+          s"Feature class '${head.getClass.getName}' doesn't implement Sizable - " +
+              "using estimated size for memory threshold calculations")
+        UnSizableFeatureIsSizable
+    }
+
+    val files = ArrayBuffer.empty[File] // tmp files we create when we exceed in-memory threshold
+    lazy val serializer = KryoFeatureSerializer(head.getFeatureType, SerializationOptions.withUserData)
+    // use ArrayList for sort-in-place of the underlying array
+    val list = new java.util.ArrayList[SimpleFeature](100)
+    list.add(head)
+    var size = sizable.sizeOf(head)
+
+    // note: we already ensured tail.hasNext == true before calling this method
+    do {
+      if (size >= threshold) {
+        // write out the sorted list to disk
+        list.sort(ordering)
+        files += Files.createTempFile(s"gm-sort-$id-", ".kryo").toFile
+        logger.trace(s"Created temp sort file '${files.last.getAbsolutePath}'")
+        WithClose(new FileOutputStream(files.last)) { os =>
```

Comment:
Is files.last 'quick'?  Or would it be better to have a 'val file = Files.createTempFile', operate on it, and then add it to 'files'?

Reply:
it's an array buffer, so it should be just an array lookup - i thought it was safer to add it directly to the `files` object but i can make it a local val if you think that's better.

Reply:
Makes sense.   Maybe check that fine detail and maybe add a comment?

Reply:
i made it a local val